File: bindings/php/src/Amber.php
<?php

declare(strict_types=1);

namespace Amber;

/** Thrown when a capture fails or an argument is rejected. */
final class CaptureException extends \RuntimeException {}

/** Output-format selectors (mirror the C ABI AMBER_FORMAT_*). */
final class Format
{
    public const HTML = 0;
    public const MHTML = 1;
    public const MARKDOWN = 2;
    public const READABLE = 3;
    public const WARC = 4;
    public const WACZ = 5;
    public const SCREENSHOT = 6;
    public const PDF = 7;
}

/**
 * Local-first web-page capture via the amber-core C ABI (PHP FFI). See
 * Plans.md (task 11.1). Run generate.sh first to stage the native library
 * (lib/libamber_core.{dylib,so}); set AMBER_LIB to point elsewhere.
 *
 *   use Amber\Amber;
 *   use Amber\Format;
 *   $md  = Amber::captureMarkdown("https://example.com");
 *   $pdf = Amber::capture("https://example.com", Format::PDF);
 *   $snap = Amber::snapshot("https://example.com", [Format::MARKDOWN, Format::PDF]);
 *   $snap->save(Format::HTML, "out", "page");
 */
final class Amber
{
    private const OK = 0;
    private const ERR_INVALID_INPUT = 1;

    private const CDEF = <<<'C'
        int amber_capture_markdown(const char *url, char **out);
        int amber_capture_readable(const char *url, char **out);
        int amber_capture(const char *url, int format, uint8_t **out, size_t *out_len);
        int amber_save(const char *url, int format, const char *dir, const char *name, char **out_path);
        typedef struct AmberSnapshot AmberSnapshot;
        int amber_snapshot(const char *url, const int *formats, size_t n_formats, AmberSnapshot **out);
        int amber_snapshot_render(const AmberSnapshot *snap, int format, uint8_t **out, size_t *out_len);
        int amber_snapshot_text(const AmberSnapshot *snap, int format, char **out);
        int amber_snapshot_save(const AmberSnapshot *snap, int format, const char *dir, const char *name, char **out_path);
        void amber_snapshot_free(AmberSnapshot *snap);
        void amber_string_free(char *s);
        void amber_bytes_free(uint8_t *ptr, size_t len);
        C;

    private static ?\FFI $ffi = null;

    /** @internal Shared FFI handle (also used by Snapshot). */
    public static function lib(): \FFI
    {
        if (self::$ffi === null) {
            $ext = \PHP_OS_FAMILY === 'Darwin' ? 'dylib' : 'so';
            $lib = getenv('AMBER_LIB') ?: \dirname(__DIR__) . "/lib/libamber_core.$ext";
            self::$ffi = \FFI::cdef(self::CDEF, $lib);
        }
        return self::$ffi;
    }

    /** @internal Map a non-zero status code to an exception. */
    public static function fail(int $rc): CaptureException
    {
        return new CaptureException($rc === self::ERR_INVALID_INPUT ? 'invalid input' : 'capture failed');
    }

    /** Capture $url and return its clean Markdown. */
    public static function captureMarkdown(string $url): string
    {
        $ffi = self::lib();
        $out = $ffi->new('char*');
        $rc = $ffi->amber_capture_markdown($url, \FFI::addr($out));
        if ($rc !== self::OK) {
            throw self::fail($rc);
        }
        try {
            return \FFI::string($out);
        } finally {
            $ffi->amber_string_free($out);
        }
    }

    /** Capture $url and return its readable plain text. */
    public static function captureReadable(string $url): string
    {
        $ffi = self::lib();
        $out = $ffi->new('char*');
        $rc = $ffi->amber_capture_readable($url, \FFI::addr($out));
        if ($rc !== self::OK) {
            throw self::fail($rc);
        }
        try {
            return \FFI::string($out);
        } finally {
            $ffi->amber_string_free($out);
        }
    }

    /**
     * Capture $url as $format and return the encoded bytes (a binary string).
     * Works for every format, including binary ones (Screenshot/PDF/MHTML/…).
     */
    public static function capture(string $url, int $format): string
    {
        $ffi = self::lib();
        $out = $ffi->new('uint8_t*');
        $len = $ffi->new('size_t');
        $rc = $ffi->amber_capture($url, $format, \FFI::addr($out), \FFI::addr($len));
        if ($rc !== self::OK) {
            throw self::fail($rc);
        }
        try {
            return \FFI::string($out, $len->cdata);
        } finally {
            $ffi->amber_bytes_free($out, $len->cdata);
        }
    }

    /**
     * Capture $url as $format, write it into $dir, and return the written path.
     * $name is the file stem (extension follows the format); null uses a default.
     */
    public static function save(string $url, int $format, string $dir, ?string $name = null): string
    {
        $ffi = self::lib();
        $out = $ffi->new('char*');
        $rc = $ffi->amber_save($url, $format, $dir, $name, \FFI::addr($out));
        if ($rc !== self::OK) {
            throw self::fail($rc);
        }
        try {
            return \FFI::string($out);
        } finally {
            $ffi->amber_string_free($out);
        }
    }

    /**
     * Capture $url once for the given $formats, returning a reusable Snapshot —
     * capture once, emit many. $formats must be non-empty.
     *
     * @param int[] $formats
     */
    public static function snapshot(string $url, array $formats): Snapshot
    {
        return Snapshot::capture($url, $formats);
    }
}

/**
 * A captured page, reusable across many output formats (Plans.md 10.1/11.1).
 * One capture serves every format with no re-fetch and no re-render. The native
 * handle is freed on close() or destruction.
 */
final class Snapshot
{
    private \FFI $ffi;
    /** @var \FFI\CData|null AmberSnapshot* */
    private $ptr;

    private function __construct(\FFI $ffi, $ptr)
    {
        $this->ffi = $ffi;
        $this->ptr = $ptr;
    }

    /**
     * Capture $url once for $formats.
     *
     * @param int[] $formats
     */
    public static function capture(string $url, array $formats): self
    {
        $ffi = Amber::lib();
        $n = \count($formats);
        $arr = $n > 0 ? $ffi->new("int[$n]") : null;
        $i = 0;
        foreach ($formats as $f) {
            $arr[$i++] = (int) $f;
        }
        $out = $ffi->new('AmberSnapshot*');
        $rc = $ffi->amber_snapshot($url, $arr, $n, \FFI::addr($out));
        if ($rc !== 0) {
            throw Amber::fail($rc);
        }
        return new self($ffi, $out);
    }

    /** Render $format from the captured page as encoded bytes. */
    public function render(int $format): string
    {
        $out = $this->ffi->new('uint8_t*');
        $len = $this->ffi->new('size_t');
        $rc = $this->ffi->amber_snapshot_render($this->ptr, $format, \FFI::addr($out), \FFI::addr($len));
        if ($rc !== 0) {
            throw Amber::fail($rc);
        }
        try {
            return \FFI::string($out, $len->cdata);
        } finally {
            $this->ffi->amber_bytes_free($out, $len->cdata);
        }
    }

    /** Render $format from the captured page as UTF-8 text. */
    public function text(int $format): string
    {
        $out = $this->ffi->new('char*');
        $rc = $this->ffi->amber_snapshot_text($this->ptr, $format, \FFI::addr($out));
        if ($rc !== 0) {
            throw Amber::fail($rc);
        }
        try {
            return \FFI::string($out);
        } finally {
            $this->ffi->amber_string_free($out);
        }
    }

    /** Write $format into $dir; returns the written path. */
    public function save(int $format, string $dir, ?string $name = null): string
    {
        $out = $this->ffi->new('char*');
        $rc = $this->ffi->amber_snapshot_save($this->ptr, $format, $dir, $name, \FFI::addr($out));
        if ($rc !== 0) {
            throw Amber::fail($rc);
        }
        try {
            return \FFI::string($out);
        } finally {
            $this->ffi->amber_string_free($out);
        }
    }

    /** The captured page's clean Markdown. */
    public function markdown(): string
    {
        return $this->text(Format::MARKDOWN);
    }

    /** The captured page's readable plain text. */
    public function readable(): string
    {
        return $this->text(Format::READABLE);
    }

    /** Release the native handle (idempotent). */
    public function close(): void
    {
        if ($this->ptr !== null) {
            $this->ffi->amber_snapshot_free($this->ptr);
            $this->ptr = null;
        }
    }

    public function __destruct()
    {
        $this->close();
    }
}

File: bindings/lua/amber.lua
--- Lua bindings for AmberHTML, a local-first web-page capture engine.
--- Uses the LuaJIT FFI over the `amber-core` C ABI (requires LuaJIT for `ffi`).
--- See Plans.md (task 11.3). Run generate.sh first to stage the native library.
---
---   local amber = require("amber")
---   local F = amber.Format
---   local md  = amber.capture_markdown("https://example.com")
---   local pdf = amber.capture("https://example.com", F.PDF)   -- binary string
---   local snap = amber.snapshot("https://example.com", { F.MARKDOWN, F.PDF })
---   print(snap:markdown())          -- one capture, many formats
---   snap:save(F.HTML, "out", "page")
---   snap:close()

local ffi = require("ffi")

ffi.cdef([[
int amber_capture_markdown(const char *url, char **out);
int amber_capture_readable(const char *url, char **out);
int amber_capture(const char *url, int format, uint8_t **out, size_t *out_len);
int amber_save(const char *url, int format, const char *dir, const char *name, char **out_path);
typedef struct AmberSnapshot AmberSnapshot;
int amber_snapshot(const char *url, const int *formats, size_t n_formats, AmberSnapshot **out);
int amber_snapshot_render(const AmberSnapshot *snap, int format, uint8_t **out, size_t *out_len);
int amber_snapshot_text(const AmberSnapshot *snap, int format, char **out);
int amber_snapshot_save(const AmberSnapshot *snap, int format, const char *dir, const char *name, char **out_path);
void amber_snapshot_free(AmberSnapshot *snap);
void amber_string_free(char *s);
void amber_bytes_free(uint8_t *ptr, size_t len);
]])

local function load_lib()
  local override = os.getenv("AMBER_LIB")
  if override and #override > 0 then
    return ffi.load(override)
  end
  local ext = (jit.os == "OSX" and "dylib") or (jit.os == "Windows" and "dll") or "so"
  -- Try the staged library (generate.sh) before the system search path.
  for _, path in ipairs({ "lib/libamber_core." .. ext, "native/libamber_core." .. ext }) do
    local ok, lib = pcall(ffi.load, path)
    if ok then return lib end
  end
  return ffi.load("amber_core") -- fall back to the system loader
end

local C = load_lib()

local OK = 0
local ERR_INVALID_INPUT = 1

local function fail(rc)
  error(rc == ERR_INVALID_INPUT and "amber: invalid input" or "amber: capture failed", 3)
end

local M = {}

--- Output-format selectors (mirror the C ABI AMBER_FORMAT_*).
M.Format = {
  HTML = 0,
  MHTML = 1,
  MARKDOWN = 2,
  READABLE = 3,
  WARC = 4,
  WACZ = 5,
  SCREENSHOT = 6,
  PDF = 7,
}

local function capture_text_c(fn, url)
  local out = ffi.new("char*[1]")
  local rc = fn(url, out)
  if rc ~= OK then fail(rc) end
  local s = ffi.string(out[0])
  C.amber_string_free(out[0])
  return s
end

--- Capture `url` and return its clean Markdown.
function M.capture_markdown(url)
  return capture_text_c(C.amber_capture_markdown, url)
end

--- Capture `url` and return its readable plain text.
function M.capture_readable(url)
  return capture_text_c(C.amber_capture_readable, url)
end

--- Capture `url` as `format` and return the encoded bytes (a binary string).
function M.capture(url, format)
  local out = ffi.new("uint8_t*[1]")
  local len = ffi.new("size_t[1]")
  local rc = C.amber_capture(url, format, out, len)
  if rc ~= OK then fail(rc) end
  local s = ffi.string(out[0], len[0])
  C.amber_bytes_free(out[0], len[0])
  return s
end

--- Capture `url` as `format`, write it into `dir`, return the written path.
--- `name` is the file stem (extension follows the format); nil uses a default.
function M.save(url, format, dir, name)
  local out = ffi.new("char*[1]")
  local rc = C.amber_save(url, format, dir, name, out)
  if rc ~= OK then fail(rc) end
  local s = ffi.string(out[0])
  C.amber_string_free(out[0])
  return s
end

--- A captured page, reusable across many output formats (Plans.md 10.1/11.3).
local Snapshot = {}
Snapshot.__index = Snapshot

--- Capture `url` once for `formats` (an array of Format values); must be non-empty.
function M.snapshot(url, formats)
  local n = #formats
  local arr = n > 0 and ffi.new("int[?]", n) or nil
  for i = 1, n do
    arr[i - 1] = formats[i]
  end
  local out = ffi.new("AmberSnapshot*[1]")
  local rc = C.amber_snapshot(url, arr, n, out)
  if rc ~= OK then fail(rc) end
  -- Auto-free on GC; close() detaches this finalizer to avoid a double free.
  local ptr = ffi.gc(out[0], C.amber_snapshot_free)
  return setmetatable({ _ptr = ptr }, Snapshot)
end

--- Render `format` from the captured page as encoded bytes.
function Snapshot:render(format)
  local out = ffi.new("uint8_t*[1]")
  local len = ffi.new("size_t[1]")
  local rc = C.amber_snapshot_render(self._ptr, format, out, len)
  if rc ~= OK then fail(rc) end
  local s = ffi.string(out[0], len[0])
  C.amber_bytes_free(out[0], len[0])
  return s
end

--- Render `format` from the captured page as UTF-8 text.
function Snapshot:text(format)
  local out = ffi.new("char*[1]")
  local rc = C.amber_snapshot_text(self._ptr, format, out)
  if rc ~= OK then fail(rc) end
  local s = ffi.string(out[0])
  C.amber_string_free(out[0])
  return s
end

--- Write `format` into `dir`; returns the written path.
function Snapshot:save(format, dir, name)
  local out = ffi.new("char*[1]")
  local rc = C.amber_snapshot_save(self._ptr, format, dir, name, out)
  if rc ~= OK then fail(rc) end
  local s = ffi.string(out[0])
  C.amber_string_free(out[0])
  return s
end

--- The captured page's clean Markdown.
function Snapshot:markdown()
  return self:text(M.Format.MARKDOWN)
end

--- The captured page's readable plain text.
function Snapshot:readable()
  return self:text(M.Format.READABLE)
end

--- Release the native handle (idempotent).
function Snapshot:close()
  if self._ptr ~= nil then
    ffi.gc(self._ptr, nil) -- detach the GC finalizer before freeing
    C.amber_snapshot_free(self._ptr)
    self._ptr = nil
  end
end

return M
